araproc/analysis/snr.py

```diff
@@ -96,6 +97,77 @@ def get_min_segmented_rms(waveform, nsegs=8):
 
     return rms
 
+def get_jackknife_rms(waveform, nSamples=50):
+
+    """
+    Calculates the RMS of a voltage trace via a sort-of jackknife technique,
+    which attempts to identify the parts of the trace which do not contain 
+    signal with the goal of using as much of the trace to estimate the RMS 
+    as possible. The RMS is then calculated using all parts of the trace which
+    are determined to not be contain signal (ie outlier voltages). The trace
+    is segmented into equal-sample segments to avoid sensitivity to the trace
+    length.
+
+    Parameters
+    ----------
+    waveform: TGraph or np.ndarray
+        A TGraph or np.ndarray of the waveform voltage.
+    nSamples: int
+        Number of samples for each segment.
+
+    Returns
+    -------
+    rms : float
+        The jackknifed RMS in same units as trace.
+    """
+    
+    if(isinstance(waveform, ROOT.TGraph)):
+      _, trace = wfu.tgraph_to_arrays(waveform)
+    elif(isinstance(waveform, np.ndarray)):
+      trace = np.copy(waveform)
+      
+      # don't be bothered by some unused dimensions
+      trace = np.squeeze(trace)
+
+      if(trace.ndim != 1):
+        raise Exception("Trace is not 1d in snr.get_jackknife_rms. Abort")
+
+    else:
+      raise Exception("Unsupported data type in snr.get_jackknife_rms. Abort")
+
+    # first split the trace into segments of nSamples 
+    # then calculate the RMS on the trace _without_ that segment (ie the "all-but RMS")
+    allRms = []
+    traceLen = len(trace)
+    for i in range(0, traceLen, nSamples):
+
+      start = i
+      end = start + nSamples
+
+      subTrace = np.concatenate([trace[:start], trace[end+1:]])
+      thisRms = np.sqrt(np.mean(subTrace**2))
+      allRms.append(thisRms)
+   
+    # segments that have outlier (ie nonthermal) voltages, will have
+    # an all-but RMS that is significantly smaller than others
+    # here we attempt to identify those outlier segments by detecting
+    # all-but RMS values which are smaller than the mean of the other
+    # all-but RMS values by more the 1 standard deviation of the other all-but RMS values 
+    # non-outlier segments are collected to calculate the final RMS
+    allRms = np.asarray(allRms)
+    smoothedTrace = np.empty(0)
```

Comment:
```suggestion
    # Identify non-outlier segments
    subMeans = np.array([np.mean(np.delete(allRms, i)) for i in range(len(allRms))])
    subStds = np.array([np.std(np.delete(allRms, i)) for i in range(len(allRms))])
    nonOutlierMask = allRms > (subMeans - subStds)

    # Combine non-outlier segments
    smoothedTrace = np.concatenate([segments[i] for i in range(len(segments)) if nonOutlierMask[i]])
```

And maybe this part can be vectorized like this?